### ISSUE TITLE
uncompyle6-2.9.8 test failures

### ISSUE THREAD

**Issue by dol-sen** [rocky/python-uncompyle6#74]
Previous release have been passing tests fine. 
...<snip>
Tue Jan  3 12:24:01 2017
Source directory:  bytecode_pypy2.7
Output directory:  /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-84_2hf_1/bytecode_pypy2.7
decompiled 22 files: 0 okay, 0 failed
Everything good, removing /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-84_2hf_1
Tue Jan  3 12:24:02 2017
Source directory:  bytecode_1.5
Output directory:  /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-84_2hf_1/bytecode_1.5
decompiled 34 files: 0 okay, 0 failed
Everything good, removing /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-84_2hf_1
/usr/bin/python3.4 test_pythonlib.py --bytecode-3.4 --verify 
Tue Jan  3 12:24:10 2017
Source directory:  bytecode_3.4
Output directory:  /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-mnghpppd/bytecode_3.4
Code differs in '.<module>.readline' at offset 43_0 [COME_FROM_LOOP] != [POP_BLOCK]

Code differs in .<module>.readline

   3       0  SETUP_LOOP           43  'to to 43'	
  11       0  SETUP_LOOP           44  'to to 44'

   4       3  LOAD_FAST                'self'	
  12       3  LOAD_FAST                'self'
           6  LOAD_ATTR                'join_lines'	           6  LOAD_ATTR                'join_lines'
           9  POP_JUMP_IF_FALSE    39  'to 39'	           9  POP_JUMP_IF_FALSE    39  'to 39'

   5      12  LOAD_FAST                'line'	
  13      12  LOAD_FAST                'line'
          15  POP_JUMP_IF_FALSE    24  'to 24'	          15  POP_JUMP_IF_FALSE    24  'to 24'

   6      18  CONTINUE              3  'to 3'	
  14      18  CONTINUE              3  'to 3'
          21  JUMP_FORWARD         24  'to to 24'	          21  JUMP_FORWARD         24  'to to 24'

   7      24  LOAD_FAST                'self'	
  15      24  LOAD_FAST                'self'
          27  POP_JUMP_IF_FALSE    39  'to 39'	          27  POP_JUMP_IF_FALSE    39  'to 39'

   8      30  CONTINUE              3  'to 3'	
  16      30  CONTINUE              3  'to 3'
          33  JUMP_ABSOLUTE        39  'to 39'	          33  JUMP_ABSOLUTE        39  'to 39'
          36  JUMP_FORWARD         39  'to to 39'	          36  JUMP_FORWARD         39  'to to 39'

  10      39  LOAD_FAST                'line'	
  17      39  LOAD_FAST                'line'
          42  RETURN_VALUE           	          42  RETURN_VALUE           
        43_0  COME_FROM_LOOP           '0'	          43  POP_BLOCK              
          43  LOAD_CONST               ''	        44_0  COME_FROM_LOOP           '0'
          46  RETURN_VALUE           	          44  LOAD_CONST               ''

decompiled 53 files: 0 okay, 0 failed, 1 verify failed
### Error Verifying 10_while.pyc
Code differs in '.<module>.readline' at offset 43_0 [COME_FROM_LOOP] != [POP_BLOCK]

Code differs in .<module>.readline

   3       0  SETUP_LOOP           43  'to to 43'	
  11       0  SETUP_LOOP           44  'to to 44'

   4       3  LOAD_FAST                'self'	
  12       3  LOAD_FAST                'self'
           6  LOAD_ATTR                'join_lines'	           6  LOAD_ATTR                'join_lines'
           9  POP_JUMP_IF_FALSE    39  'to 39'	           9  POP_JUMP_IF_FALSE    39  'to 39'

   5      12  LOAD_FAST                'line'	
  13      12  LOAD_FAST                'line'
          15  POP_JUMP_IF_FALSE    24  'to 24'	          15  POP_JUMP_IF_FALSE    24  'to 24'

   6      18  CONTINUE              3  'to 3'	
  14      18  CONTINUE              3  'to 3'
          21  JUMP_FORWARD         24  'to to 24'	          21  JUMP_FORWARD         24  'to to 24'

   7      24  LOAD_FAST                'self'	
  15      24  LOAD_FAST                'self'
          27  POP_JUMP_IF_FALSE    39  'to 39'	          27  POP_JUMP_IF_FALSE    39  'to 39'

   8      30  CONTINUE              3  'to 3'	
  16      30  CONTINUE              3  'to 3'
          33  JUMP_ABSOLUTE        39  'to 39'	          33  JUMP_ABSOLUTE        39  'to 39'
          36  JUMP_FORWARD         39  'to to 39'	          36  JUMP_FORWARD         39  'to to 39'

  10      39  LOAD_FAST                'line'	
  17      39  LOAD_FAST                'line'
          42  RETURN_VALUE           	          42  RETURN_VALUE           
        43_0  COME_FROM_LOOP           '0'	          43  POP_BLOCK              
          43  LOAD_CONST               ''	        44_0  COME_FROM_LOOP           '0'
          46  RETURN_VALUE           	          44  LOAD_CONST               ''

make[2]: *** [Makefile:43: check-3.4] Error 3
make[2]: Leaving directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8/test'
make[1]: *** [Makefile:32: check-3.4] Error 2
make[1]: Leaving directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8'
make: *** [Makefile:23: check] Error 2
 * ERROR: dev-python/uncompyle6-2.9.8::gentoo failed (test phase):
 *   emake failed
 *

**Comment by rocky** [comment]
I tried to reproduce this and can't. On further investigation of the code, it looks like the additional `POP_BLOCK` instruction at offset 43 is generated from a Python 3.5 compiler, rather than a Python 3.4 compiler. 

You can probably verify this yourself. Start with the source

```
def readline(self, line):
    while 1:
        if self.join_lines:
            if line:
                continue
            if self:
                continue
        return line
while __name__ != '__main__':
    b = 4
```

Compile that under Python 3.4 (this source is also given in [`test/simple-source/looping/10_while.py`](https://github.com/rocky/python-uncompyle6/blob/master/test/simple_source/looping/10_while.py) and you should see the instructions on the left and no `POP_BLOCK`. Now if you compile it under Python 3.5 the `POP_BLOCK` will appear. 

Given this, a hypothesis is that somehow `uncompyle6` is fooled into thinking that it is running Python 3.4 when in fact it is running 3.5. The actual `compile()` command is done in another Python project `xdis`. See https://github.com/rocky/python-xdis/blob/master/xdis/load.py#L61  So put a `print(PYTHON_VERSION)` before that line to make sure you are compiling with Python 3.4

**Comment by dol-sen** [comment]
No, it isn't crossing the py3.4, py3.5 boundaries.  I added:

    print("****", PYTHON_VERSION)
to both python3.4's and python3.5's xdis/load.py.

I will point out that the tests passed in this same install run for py2.7.  I'll disable the install to the others and just install to 3.5 now to see if it passes there.  My system is currently set to install to all 3 pythons: 2.7, 3.4, 3.5 site-packages, but once it fails in 3.4, it quits, so had not run the tests in 3.5 yet.


Here is the output from that run:

...
Installed /var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8-python3_4/test/lib/uncompyle6-2.9.8-py3.4.egg
Processing dependencies for uncompyle6==2.9.8
Finished processing dependencies for uncompyle6==2.9.8
make -j16 -l16 check 
make[1]: Entering directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8'
make -C pytest check
make[2]: Entering directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8/pytest'
py.test
================================================================================ test session starts =================================================================================
platform linux -- Python 3.4.5, pytest-3.0.5, py-1.4.31, pluggy-0.4.0
rootdir: /var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8, inifile: 
plugins: xdist-1.15.0, virtualenv-1.2.5, timeout-1.2.0, shutil-1.2.4, pep8-1.0.6, cov-2.3.1, hypothesis-3.6.1
collected 12 items

test_deparse.py .
test_disasm.py ..
test_docstring.py .
test_fjt.py .
test_fstring.py ssss
test_grammar.py ..
test_single_compile.py .

======================================================================== 8 passed, 4 skipped in 0.41 seconds =========================================================================
make[2]: Leaving directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8/pytest'
make -C test check-3.4
make[2]: Entering directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8/test'
/usr/bin/python3.4 test_pythonlib.py --bytecode-3.0 \
       --bytecode-3.1 --bytecode-3.2 --bytecode-3.3 \
       --bytecode-3.4 --bytecode-3.5 --bytecode-pypy3.2
/usr/bin/python3.4 test_pythonlib.py --ok-3.4 --verify 
/usr/bin/python3.4 test_pythonlib.py --ok-2.7 --verify 
Wed Jan  4 09:01:39 2017
Source directory:  bytecode_3.0
...
Everything good, removing /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-graz3t3_
/usr/bin/python3.4 test_pythonlib.py --bytecode-3.4 --verify 
Wed Jan  4 09:02:20 2017
Source directory:  bytecode_3.4
Output directory:  /var/tmp/portage/dev-python/uncompyle6-2.9.8/temp/py-dis-ijt3u26j/bytecode_3.4
**** 3.4
**** 3.4ully decompiled file
**** 3.4ed 2 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 3 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 4 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 5 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 6 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 7 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 8 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 9 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 10 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 11 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 12 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 13 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 14 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 15 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 16 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 17 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 18 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 19 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 20 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 21 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 22 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 23 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 24 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 25 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 26 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 27 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 28 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 29 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 30 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 31 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 32 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 33 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 34 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 35 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 36 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 37 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 38 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 39 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 40 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 41 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 42 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 43 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 44 files: 0 okay, 0 failed, 0 verify failed
**** 3.4ed 45 files: 0 okay, 0 failed, 0 verify failed
Code differs in '.<module>.readline' at offset 43_0 [COME_FROM_LOOP] != [POP_BLOCK]

Code differs in .<module>.readline

   3       0  SETUP_LOOP           43  'to to 43'	
  11       0  SETUP_LOOP           44  'to to 44'

   4       3  LOAD_FAST                'self'	
  12       3  LOAD_FAST                'self'
           6  LOAD_ATTR                'join_lines'	           6  LOAD_ATTR                'join_lines'
           9  POP_JUMP_IF_FALSE    39  'to 39'	           9  POP_JUMP_IF_FALSE    39  'to 39'

   5      12  LOAD_FAST                'line'	
  13      12  LOAD_FAST                'line'
          15  POP_JUMP_IF_FALSE    24  'to 24'	          15  POP_JUMP_IF_FALSE    24  'to 24'

   6      18  CONTINUE              3  'to 3'	
  14      18  CONTINUE              3  'to 3'
          21  JUMP_FORWARD         24  'to to 24'	          21  JUMP_FORWARD         24  'to to 24'

   7      24  LOAD_FAST                'self'	
  15      24  LOAD_FAST                'self'
          27  POP_JUMP_IF_FALSE    39  'to 39'	          27  POP_JUMP_IF_FALSE    39  'to 39'

   8      30  CONTINUE              3  'to 3'	
  16      30  CONTINUE              3  'to 3'
          33  JUMP_ABSOLUTE        39  'to 39'	          33  JUMP_ABSOLUTE        39  'to 39'
          36  JUMP_FORWARD         39  'to to 39'	          36  JUMP_FORWARD         39  'to to 39'

  10      39  LOAD_FAST                'line'	
  17      39  LOAD_FAST                'line'
          42  RETURN_VALUE           	          42  RETURN_VALUE           
        43_0  COME_FROM_LOOP           '0'	          43  POP_BLOCK              
          43  LOAD_CONST               ''	        44_0  COME_FROM_LOOP           '0'
          46  RETURN_VALUE           	          44  LOAD_CONST               ''

**** 3.4ed 46 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 47 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 48 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 49 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 50 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 51 files: 0 okay, 0 failed, 1 verify failed
**** 3.4ed 52 files: 0 okay, 0 failed, 1 verify failed
decompiled 53 files: 0 okay, 0 failed, 1 verify failed
### Error Verifying 10_while.pyc
Code differs in '.<module>.readline' at offset 43_0 [COME_FROM_LOOP] != [POP_BLOCK]

Code differs in .<module>.readline

   3       0  SETUP_LOOP           43  'to to 43'	
  11       0  SETUP_LOOP           44  'to to 44'

   4       3  LOAD_FAST                'self'	
  12       3  LOAD_FAST                'self'
           6  LOAD_ATTR                'join_lines'	           6  LOAD_ATTR                'join_lines'
           9  POP_JUMP_IF_FALSE    39  'to 39'	           9  POP_JUMP_IF_FALSE    39  'to 39'

   5      12  LOAD_FAST                'line'	
  13      12  LOAD_FAST                'line'
          15  POP_JUMP_IF_FALSE    24  'to 24'	          15  POP_JUMP_IF_FALSE    24  'to 24'

   6      18  CONTINUE              3  'to 3'	
  14      18  CONTINUE              3  'to 3'
          21  JUMP_FORWARD         24  'to to 24'	          21  JUMP_FORWARD         24  'to to 24'

   7      24  LOAD_FAST                'self'	
  15      24  LOAD_FAST                'self'
          27  POP_JUMP_IF_FALSE    39  'to 39'	          27  POP_JUMP_IF_FALSE    39  'to 39'

   8      30  CONTINUE              3  'to 3'	
  16      30  CONTINUE              3  'to 3'
          33  JUMP_ABSOLUTE        39  'to 39'	          33  JUMP_ABSOLUTE        39  'to 39'
          36  JUMP_FORWARD         39  'to to 39'	          36  JUMP_FORWARD         39  'to to 39'

  10      39  LOAD_FAST                'line'	
  17      39  LOAD_FAST                'line'
          42  RETURN_VALUE           	          42  RETURN_VALUE           
        43_0  COME_FROM_LOOP           '0'	          43  POP_BLOCK              
          43  LOAD_CONST               ''	        44_0  COME_FROM_LOOP           '0'
          46  RETURN_VALUE           	          44  LOAD_CONST               ''

make[2]: *** [Makefile:43: check-3.4] Error 3
make[2]: Leaving directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8/test'
make[1]: *** [Makefile:32: check-3.4] Error 2
make[1]: Leaving directory '/var/tmp/portage/dev-python/uncompyle6-2.9.8/work/uncompyle6-2.9.8'
make: *** [Makefile:23: check] Error 2

**Comment by dol-sen** [comment]
Indeed, it is only failing for python3.4.  Confirmed it passed all tests in 3.5.

I wonder if it is the py3.4 version I have if it is different that the one you've tested against.

I'll ask if any 3.5 patches have been backported by us (I doubt it) just in case.

**Comment by rocky** [comment]
I have 3.4.2 installed and the changes to the magic numbers listed in xdis's magics.py wouldn't account for the difference you encounter. Compiling with 3.5 definitely would. 

So some things to try to narrow. Have the code at https://github.com/rocky/python-uncompyle6/blob/master/uncompyle6/verify.py#L102 exit so you can inspect the produced bytecode which will I believe be in `/tmp` somwhere. Then see what magic number is stored in the `pyc` file.  `pydisasm` will show you that.  I bet the magic will be associated with 3.5.  If it is, then try removing _any_ Python 3.5 interpreter on the system you are testing with.

It is I think a well known problem that different versions of python get can run. On my system I have a `/usr/lib/python3/dist-packages` so that's one place there could be a bleeding of different versions coming through.

The bottom line here though is that the uncompyle6 decompilation code is not failing and you can check that by looking at the source code it produced and compare that with the source file that is in the distribution. What _is_ failing is the strong verification part.  And that is failing because the bytecode it compares against looks like it was produced by a different kind of Python interpreter.

**Comment by rocky** [comment]
One other suspicious thing. The double "to" in `3 0 SETUP_LOOP 43 'to to 43'` listed above was fixed recently and I think went into the 2.9.9 release. See commit  39814fab and https://github.com/rocky/python-uncompyle6/blob/master/uncompyle6/scanners/tok.py#L70
